### PR TITLE
Handle failed SaveAsync operations

### LINKED ISSUE #31548: Unexpected Null RecoverableTextAndVersion in unit tests
**Version Used**: `2.11.0-beta2-63527-06`

**Steps to Reproduce**:

I'm hitting a `Contract.Fail` inside some Roslyn types in the unit tests for Razor Find All References. This is sporadic and doesn't have a consistent repro but I've hit it enough by now that it's becoming troublesome. 

Example of a test that can sometimes fail: https://github.com/aspnet/Razor/blob/6c1bee1940e94387d96b03ed03837ad232a82099/test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/RazorExcerptServiceTest.cs#L507

Stack Trace:

```
  at Microsoft.CodeAnalysis.DocumentState.GetSyntaxTreeAsync(CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs:line 631
   at Microsoft.CodeAnalysis.Document.GetSyntaxRootAsync(CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Workspace/Solution/Document.cs:line 204
   at Microsoft.CodeAnalysis.Shared.Extensions.DocumentExtensions.GetSemanticModelForSpanAsync(Document document, TextSpan span, CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Shared/Extensions/DocumentExtensions.cs:line 58
   at Microsoft.CodeAnalysis.Classification.Classifier.GetClassifiedSpansAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Classification/Classifier.cs:line 23
   at Microsoft.CodeAnalysis.Razor.RazorDocumentExcerptService.ClassifyPreviewAsync(SourceText primaryText, TextSpan excerptSpan, Document secondaryDocument, IReadOnlyList`1 mappings, CancellationToken cancellationToken) in /_/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/RazorDocumentExcerptService.cs:line 175
   at Microsoft.CodeAnalysis.Razor.RazorDocumentExcerptService.TryGetExcerptInternalAsync(Document document, TextSpan span, ExcerptModeInternal mode, CancellationToken cancellationToken) in /_/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/RazorDocumentExcerptService.cs:line 81
   at Microsoft.CodeAnalysis.Razor.RazorExcerptServiceTest.TryGetExcerptInternalAsync_MultiLine_Boundaries_CanClassifyCSharp() in /_/src/Razor/test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/RazorExcerptServiceTest.cs:line 408
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
   at Microsoft.CodeAnalysis.Host.RecoverableWeakValueSource`1.GetValueAsync(CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Utilities/ValuesSources/RecoverableWeakValueSource.cs:line 105
   at Microsoft.CodeAnalysis.RecoverableTextAndVersion.GetValueAsync(CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:line 98
   at Microsoft.CodeAnalysis.DocumentState.FullyParseTreeAsync(ValueSource`1 newTextSource, ProjectId cacheKey, String filePath, ParseOptions options, HostLanguageServices languageServices, SolutionServices solutionServices, PreservationMode mode, CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs:line 130
----- Inner Stack Trace -----
   at Roslyn.Utilities.Contract.Fail(String message) in /_/src/Workspaces/Core/Portable/Utilities/Contract.cs:line 52
   at Roslyn.Utilities.Contract.ThrowIfNull[T](T value, String message) in /_/src/Workspaces/Core/Portable/Utilities/Contract.cs:line 21
   at Microsoft.CodeAnalysis.RecoverableTextAndVersion.RecoverableText.RecoverAsync(CancellationToken cancellationToken) in /_/src/Workspaces/Core/Portable/Workspace/Solution/RecoverableTextAndVersion.cs:line 129
```

### LINKED ISSUE #42178: RecoverableText.RecoverAsync fails null check on non-Windows platforms
On Linux and MacOS platforms `RecoverableText.RecoverAsync` can fail a null check.

```csharp
protected override async Task<SourceText> RecoverAsync(CancellationToken cancellationToken)
{
    Contract.ThrowIfNull(_storage);

    using (Logger.LogBlock(FunctionId.Workspace_Recoverable_RecoverTextAsync, _parent._filePath, cancellationToken))
    {
        return await _storage.ReadTextAsync(cancellationToken).ConfigureAwait(false);
    }
}
```

```console
jorobich@jorobich-vm1:~/Source/format$ dotnet test
Test run for /home/jorobich/Source/format/artifacts/bin/dotnet-format.UnitTests/Debug/netcoreapp3.1/dotnet-format.UnitTests.dll(.NETCoreApp,Version=v3.1)
Microsoft (R) Test Execution Command Line Tool Version 16.3.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...

A total of 1 test files matched the specified pattern.
[xUnit.net 00:00:05.88]     Microsoft.CodeAnalysis.Tools.Tests.CodeFormatterTests.NoFilesFormattedInFormattedProject [FAIL]                                                                                                                                                                                                                
  X Microsoft.CodeAnalysis.Tools.Tests.CodeFormatterTests.NoFilesFormattedInFormattedProject [182ms]                                                                                                                                                                                                                                       
  Error Message:
   System.AggregateException : One or more errors occurred. (Unexpected Null)
---- System.InvalidOperationException : Unexpected Null
  Stack Trace:
     at Microsoft.CodeAnalysis.Host.WeaklyCachedRecoverableValueSource`1.GetValueAsync(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.RecoverableTextAndVersion.GetValueAsync(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.TextDocumentState.GetTextAndVersionAsync(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.TextDocumentState.GetTextAsync(CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Tools.CodeFormatter.<>c__DisplayClass12_0.<<GetAnalyzerConfigSet>b__0>d.MoveNext() in /home/jorobich/Source/format/src/CodeFormatter.cs:line 382
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.CodeAnalysis.Tools.CodeFormatter.GetAnalyzerConfigSet(Project project, CancellationToken cancellationToken) in /home/jorobich/Source/format/src/CodeFormatter.cs:line 381
   at Microsoft.CodeAnalysis.Tools.CodeFormatter.GetAnalyzerConfigOptions(Document document, CancellationToken cancellationToken) in /home/jorobich/Source/format/src/CodeFormatter.cs:line 364
   at Microsoft.CodeAnalysis.Tools.CodeFormatter.GetDocumentAndOptions(Project project, DocumentId documentId, Matcher fileMatcher, CancellationToken cancellationToken) in /home/jorobich/Source/format/src/CodeFormatter.cs:line 328
   at Microsoft.CodeAnalysis.Tools.CodeFormatter.DetermineFormattableFiles(Solution solution, String projectPath, Matcher fileMatcher, ILogger logger, CancellationToken cancellationToken) in /home/jorobich/Source/format/src/CodeFormatter.cs:line 284
   at Microsoft.CodeAnalysis.Tools.CodeFormatter.FormatWorkspaceAsync(FormatOptions options, ILogger logger, CancellationToken cancellationToken) in /home/jorobich/Source/format/src/CodeFormatter.cs:line 65
   at Microsoft.CodeAnalysis.Tools.Tests.CodeFormatterTests.TestFormatWorkspaceAsync(String workspaceFilePath, IEnumerable`1 include, IEnumerable`1 exclude, Int32 expectedExitCode, Int32 expectedFilesFormatted, Int32 expectedFileCount) in /home/jorobich/Source/format/tests/CodeFormatterTests.cs:line 367
   at Microsoft.CodeAnalysis.Tools.Tests.CodeFormatterTests.NoFilesFormattedInFormattedProject() in /home/jorobich/Source/format/tests/CodeFormatterTests.cs:line 44
--- End of stack trace from previous location where exception was thrown ---
----- Inner Stack Trace -----
   at Roslyn.Utilities.Contract.Fail(String message)
   at Roslyn.Utilities.Contract.ThrowIfNull[T](T value, String message)
   at Microsoft.CodeAnalysis.RecoverableTextAndVersion.RecoverableText.RecoverAsync(CancellationToken cancellationToken)
                                                                                                                                                                                                                                                                                                                                           
Test Run Failed.
Total tests: 83
     Passed: 82
     Failed: 1
 Total time: 8.5459 Seconds
```

When debugging on Linux we see several PlatformNotSupportedExceptions are thrown, in particular, for System.IO.MemoryMappedFiles.
![image](https://user-images.githubusercontent.com/611219/75933434-dbf51500-5e2e-11ea-997f-4f14a075792f.png)

The TemporaryStorageService used by RecoverableText uses MemoryMappedFiles which explains this failure.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #31548
Fixes #42178

### CODE REVIEW

**Review by Youssef1313 — COMMENTED** [review]
https://github.com/dotnet/roslyn/blob/afd10305a37c0ffb2cfb2c2d8446154c68cfa87a/src/Workspaces/Core/Portable/TemporaryStorage/TemporaryStorageServiceFactory.cs#L32-L42

Could you update this code as appropriate? It mentions one of the issues being fixed in this PR.

**Comment by sharwell** [comment]
@Youssef1313 Thanks for that find. I had only searched for one of the two issues; for this one it turns out the link was just incorrectly added, so I have redirected it to the true limiting issue.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ValuesSources/WeaklyCachedRecoverableValueSource.cs [hunk 1]
```diff
@@ -128,7 +128,11 @@ private void ResetRecoverySource(Task saveTask, T instance)
             {
                 using (Gate.DisposableWait(CancellationToken.None))
                 {
-                    _recoverySource = new AsyncLazy<T>(RecoverAsync, Recover, cacheResult: false);
+                    // Only assume the instance is saved if the saveTask completed successfully. If the save did not
+                    // complete, we can still rely on a constant value source to provide the instance.
+                    _recoverySource = saveTask.Status == TaskStatus.RanToCompletion
+                        ? new AsyncLazy<T>(RecoverAsync, Recover, cacheResult: false)
+                        : new ConstantValueSource<T>(instance);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ValuesSources/WeaklyCachedRecoverableValueSource.cs:135]: this makes sense.  though i worry about waht is breaking teh saveTask.  do we have insight on that?  Can we report/log failures of that as well so we can track down any other necessary fixes?

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ValuesSources/WeaklyCachedRecoverableValueSource.cs (hunk 1 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/ValuesSources/WeaklyCachedRecoverableValueSource.cs:135]: We already do log NFE for failure of `saveTask`. @ryzngard could look up telemetry for it.

**Pull request merged**